src/callbacks/ltfb.cpp

```diff
@@ -34,283 +33,551 @@
+
+class CheckpointBlob final
+  : public Cloneable<CheckpointBlob, LTFBCommunicationAlgorithm>
+{
+  using BaseType = Cloneable<CheckpointBlob, LTFBCommunicationAlgorithm>;
+public:
+  CheckpointBlob(std::set<std::string> const& weights_names)
+    : BaseType(weights_names)
+  {}
+  CheckpointBlob(std::set<std::string>&& weights_names)
+    : BaseType(std::move(weights_names))
+  {}
+  void exchange_models(model& m,
+                       El::Int partner_trainer,
+                       El::Int /*step*/) const final
+  {
+    auto&& comm = *m.get_comm();
+    // Keep track of weights that shouldn't be exchanged
+    std::unordered_map<std::string, std::unique_ptr<weights>> restore_weights;
+    if (this->has_weights_names()) {
+      auto const& weights_names = this->weights_names();
+      for (auto w : m.get_weights()) {
+        if (weights_names.find(w->get_name()) == weights_names.cend())
+        {
+          using TensorDataType = DataType;
+          using WeightsType = data_type_weights<TensorDataType>;
+          restore_weights[w->get_name()]
+            = make_unique<WeightsType>(dynamic_cast<WeightsType&>(*w));
+        }
+      }
+    }
 
-  // Restore weights that shouldn't be exchanged
-  if (!restore_weights.empty()) {
-    for (auto w : m.get_weights()) {
-      if (restore_weights.count(w->get_name()) > 0) {
-        using TensorDataType = DataType;
-        using WeightsType = data_type_weights<TensorDataType>;
-        dynamic_cast<WeightsType&>(*w)
-          = dynamic_cast<WeightsType&>(*restore_weights[w->get_name()]);
+    // Save model checkpoint
+    std::ostringstream oss;
+    {
+      RootedBinaryOutputArchive ar(oss, comm.get_trainer_grid());
+      comm.trainer_barrier();
+      ar(m);
+    }
+
+    // sure, why not
+    comm.trainer_barrier();
+
+    // Synchronize with partner trainer
+    std::string save_model_ckpt = oss.str(), load_model_ckpt;
+    if (comm.am_trainer_master())
+    {
+      std::size_t save_size=save_model_ckpt.size(), load_size=0;
+      comm.sendrecv(&save_size, 1, partner_trainer, 0,
+                    &load_size, 1, partner_trainer, 0,
+                    El::SyncInfo<El::Device::CPU>{});
+      load_model_ckpt.resize(load_size);
+
+      while (save_size && load_size)
+      {
+        // Get the max blk size
+        auto constexpr max_blk_size = std::numeric_limits<int>::max();
+        std::size_t constexpr max_blk_size_size_t = max_blk_size;
+
+        int this_blk_send_size =
+          (save_size > max_blk_size_size_t ? max_blk_size : save_size);
+        int this_blk_recv_size =
+          (load_size > max_blk_size_size_t ? max_blk_size : load_size);
+        comm.sendrecv(
+          reinterpret_cast<El::byte const*>(save_model_ckpt.data()),
+          this_blk_send_size, partner_trainer, 0,
+          reinterpret_cast<El::byte*>(load_model_ckpt.data()),
+          this_blk_recv_size, partner_trainer, 0,
+          El::SyncInfo<El::Device::CPU>{});
```

Comment:
```suggestion
        comm.sendrecv(
          reinterpret_cast<El::byte const*>(save_model_ckpt.data() + save_model_ckpt.size() - send_size),
          this_blk_send_size, partner_trainer, 0,
          reinterpret_cast<El::byte*>(load_model_ckpt.data() + load_model_ckpt.size() - load_size),
          this_blk_recv_size, partner_trainer, 0,
          El::SyncInfo<El::Device::CPU>{});
```
Also, does sendrecv gracefully handle the case where one of the buffers is empty? If we want to be paranoid, we could replace it with a send or recv.

Reply:
Oh yeah. One also has to bump the pointers?? Epic brain fail. Solving this in a different way.

Reply:
Based on a test with MVAPICH2, it handles the case where one of the buffers is empty. I did test this idea and it worked for the cases (a) both buffers same size, (b) buffers off by one block, (c) one buffer completely empty. Apparently I just failed to copy-paste from my prototype code... A lot...